Documentation/designs/agent-deployment-pattern-sidecar.md

```diff
@@ -0,0 +1,116 @@
+
+Also for the sake of speeding up implementation, this document won't focus on several new Deployment patterns, but only one: Sidecar Injection.
+
+Looking at the traditional deployment model, we have a single Prometheus (or an HA setup) per cluster or namespace, responsible for scraping all containers under their scope. Prometheus operator relies on `ServiceMonitor`, `PodMonitor`, and `Probe` CRs to configure Prometheus, which will eventually use Kubernetes service-discovery to find endpoints that need to be scraped.
+
+Depending on the Cluster's scale and how often Prometheus hits Kubernetes API, Prometheus service discovery can increase the load on the API significantly and affect the overall functionality of said cluster.
+
+Another problem is that one or more containers can be updated to a problematic version that causes a [Cardinality Spike](https://grafana.com/blog/2022/02/15/what-are-cardinality-spikes-and-why-do-they-matter/). Depending on the proportion of the spike, it is possible that a container could single-handedly crash the monitoring system of the whole cluster.
+
+![Traditional Deployment Pattern](assets/agent-deployment-pattern-sidecar/traditional-deployment-pattern.png "Traditional Deployment Pattern").
+
+## Proposal
+
+This document proposes a new deployment model where Prometheus-Operator injects Prometheus agents as a sidecar container (and Prometheus config reloader) to pods that needs to be scrapped. With a sidecar, we tackle both problems mentioned above:
+
+> * Load on Kubernetes API won't exist since it's not needed anymore. Prometheus will scrape containers from the same pod through their shared network interface and scrape configuration can be declared via pod annotations.
+> * A sudden cardinality spike will not affect the whole monitoring system. In a worst-case scenario, it will fail a single pod.
+
+A common pattern used with Prometheus's Kubernetes service discovery is the [usage of annotation to declaratively tell Prometheus which endpoints need to be scraped](https://www.acagroup.be/en/blog/auto-discovery-of-kubernetes-endpoint-services-prometheus/). From a [code search at Github](https://github.com/search?q=prometheus.io%2Fscrape%3A+%22true%22&type=code) for `prometheus.io/scrape: "true"`, we can tell that this approach has good adoption already. To not conflict with the already commonly used annotation, we can start with our own, but with a very similar approach.
```

Comment:
I kind of see this as the "not using the operator" pattern.

This makes me wonder what it would take to allow configuring the sidecar injection with PodMonitors? In other words, inject the sidecar into any (new?) pod that matches any PodMonitor. In principle this could even extend to ServiceMonitors, by resolving the label matchers on the service.

A big benefit of this would be that it makes the choice of deployment model independent from the choice of configuration. As an observability team, one could keep consuming the same CRs (which may be owned by service teams), while transparently moving between agent deployment models, or even running multiple variants in parallel. The annotations approach would require a large coordinated effort across the organization to make sure the annotations are there and represent the same semantic intentions.

Reply:
An open question for me would be what to do when a monitor changes after a pod has started. I don't know how much flexibility Kubernetes has around that today. Theoretically this is also a problem with annotations, although in practice it's easier to say "don't do that" when it's so closely tied to the pod.

Reply:
The conceptual shift from the way you write in the implementation is that I don't see PodMonitor as _configuring Prometheus_, I see it as declaring the intention to monitor a pod. In that sense, it doesn't matter which objects or files need to be changed to fulfill that intention.

Reply:
I like the idea of using `PodMonitor` for simple deployments, would there be benefits in using annotations over changing `PodMonitor`? 
I can think of the following: 
* having the same Agent over multiple sets of pods and not an Agent per selector.
* having a platform team provide Agents for different use-cases and users select the ones they need and thus removing the burden of configuring the agent from the users

Reply:
Oh wow, I don't know how I didn't think of that before. Indeed it sounds much more aligned with the way we do things right now. 

We should be able to find and inject pods from Pod/ServiceMonitorSelectors and even inject blockbox-exproters with ProbeSelectors.

The only one that will probably be odd is the upcoming ScrapeConfigSelector, but we can defer the decision with this one to a future iteraction

Reply:
My first thought was "eh just ignore it" but my second thought was "hold on this could actually be really useful". The scrape config for a sidecar would be fairly static (mostly targeting localhost) so it would be quite easy to manage and at the same time offer a good amount of flexibility. One thing I am wondering about is how one could provide "pod metadata" to the sidecar when supplying a custom scrape config, but maybe that is sufficiently covered by external labels.

Reply:
> users select the ones they need and thus removing the burden of configuring the agent from the users

In first approximation, I don't think they should even care at this level. To me, the ideal model is that users provide a PodMonitor that describes how their pod exposes metrics, and the platform team does ✨ magic ✨ to make those metrics appear in the query interface. If the platform team wants to move from a central statefulset to a daemonset to a sidecar, nothing should change to the users.

The world isn't ideal, so sometimes you do need to give a little bit more control – this is still possible by labeling the PodMonitors and restricting the podMonitorSelectors accordingly. You can make this explicit (label `agentModel=sidecar`), or better, intention based. For example, the pod monitor could be labeled `size=small` for pods that are small (where a sidecar would add significant overhead), but the platform team can decide whether to eat that cost dor uniformity, or split this class of PodMonitors out to be covered by a statefulset.

Reply:
> My first thought was "eh just ignore it" but my second thought was "hold on this could actually be really useful". The scrape config for a sidecar would be fairly static (mostly targeting localhost) so it would be quite easy to manage and at the same time offer a good amount of flexibility. One thing I am wondering about is how one could provide "pod metadata" to the sidecar when supplying a custom scrape config, but maybe that is sufficiently covered by external labels.

Pod/Service monitors have labelSelector for pod discovery, something that ScrapeConfig doesn't have. If someone is relying entirely on the ScrapeConfig CRD to inject the agent, I believe we'll need to pod annotations that were in the original proposal